lib/tree.rb

```diff
@@ -2,63 +2,130 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(h) where h is the max height of the binary search tree.
+  # In a best case scenario, time complexity would be O(logn)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else
+      find_helper(current_node.right, key)
+    end
+  end
+
+  # Time Complexity: If tree is balanced, it would be O(logn) n being the size of the tree
+  # In a worst case scenario, the time complexity would be O(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(h) h is the height of the tree
+  # Space Complexity: O(h) h is the height of the tree
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(h) h is the height of the tree
+  # Space Complexity: O(h) h is the height of the tree
   def preorder
```

Comment:
Since you're visiting each node and creating a list containing all node values... the time/space complexity is O(n).